core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java

```diff
@@ -0,0 +1,285 @@
+     * @throws NullPointerException If {@code pollerOptions pollOperation} is {@code null}.
+     */
+    public Poller(Duration pollInterval, Function<PollResponse<T>, Mono<PollResponse<T>>> pollOperation, Consumer<Poller> cancelOperation) {
+        this(pollInterval, pollOperation);
+        this.cancelOperation = cancelOperation;
+    }
+
+    /**
+     * Calls Cancel Operation function if provided.
+     * It will only call cancelOperation if {@link com.azure.core.util.polling.PollResponse.OperationStatus} is IN_PROGRESS otherwise it does nothing.
+     *
+     * @throws UnsupportedOperationException when cancel operation is not provided.
+     */
+    public void cancelOperation() throws UnsupportedOperationException {
+        if (this.cancelOperation == null) {
+            throw new UnsupportedOperationException("Cancel operation is not supported on this service/resource.");
+        }
+
+        // We can not cancel an operation if it was never started
+        // It only make sense to call cancel operation if current status IN_PROGRESS.
+        if (this.pollResponse != null && this.pollResponse.getStatus() != PollResponse.OperationStatus.IN_PROGRESS) {
+            return;
+        }
+        //Time to call cancel
+        this.cancelOperation.accept(this);
+    }
+
+    /**
+     * Enable user to subscribe to {@link Flux} and listen to every {@link PollResponse} asynchronously.
+     * The user will start receiving {@link PollResponse} when client subscribe to this Flux.
+     * The poller could still have its own auto polling in action unless user has turned off
+     * auto polling.
+     *
+     * @return poll response as Flux that can be subscribed.
+     */
+    public Flux<PollResponse<T>> getObserver() {
+        return this.fluxHandle;
+    }
+
+    /**
+     * Enable user to take control of polling and trigger manual poll operation. It will perform one call to poll operation.
+     * This will not turn off auto polling.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p><strong>Manual Polling</strong></p>
+
+     * {@codesnippet com.azure.core.util.polling.poller.poll.indepth}
+     *
+     * @return a Mono of {@link PollResponse} This will call poll operation once. The {@link Mono} returned here could be subscribed
+     * for receiving {@link PollResponse} in async manner.
+     */
+    public Mono<PollResponse<T>> poll() {
+        return this.pollOperation.apply(this.pollResponse)
+            .doOnEach(pollResponseSignal -> {
+                if (pollResponseSignal.get() != null) {
+                    this.pollResponse = pollResponseSignal.get();
+                }
+            });
+    }
+
+    /**
+     * Blocks execution and wait for polling to complete. The polling is considered finished based on status defined in {@link com.azure.core.util.polling.PollResponse.OperationStatus}.
+     * Auto polling must be turned on for poller to continuously poll in background.
```

Comment:
This line about auto-polling must be on needs reconciling. Rather than state this to the user, why not just silently re-enable auto-polling, if that is necessary?

Reply:
"silently re-enable auto-polling" : This would be thought as side effect of calling this function.  Is this considered right in API design ? I have heard opinion both ways. But, in this case, I am on side of "lets enable auto-polling if it is disable" because the fact that user is calling this function, user want to poll.  Would you agree ?

Reply:
You alternate option is to create a new flux that does not update the poll operation state.

Reply:
I think having one additional Flux complicate the code a bit and will confuse reader a bit. I will  go with "lets enable auto-polling if it is disable"  ?